### PR TITLE
QFieldCloud integration

### PR DESCRIPTION
**Pull request by suricactus**
image

### CODE REVIEW

**Review by signedav — COMMENTED** [review]
It does look good considering merging into the master.
The code from the cloud I haven't checked that deeply.

File: test/test.pro [hunk 1]
```diff
@@ -0,0 +1 @@
+
```
[the remark this answers is not in the document]
Comment (suricactus) [reply test/test.pro:1]: To be removed.

[the remark this answers is not in the document]
Comment (suricactus) [reply test/test.pro:1]: To be removed. Ghosts from the past.

File: qfield.pri [hunk 2]
```diff
@@ -0,0 +1 @@
+
```
[the remark this answers is not in the document]
Comment (suricactus) [reply qfield.pri:1]: To be removed. Ghosts from the past.

File: src/core/core.pro [hunk 3]
```diff
@@ -0,0 +1 @@
+
```
[the remark this answers is not in the document]
Comment (suricactus) [reply src/core/core.pro:1]: To be removed. Ghosts from the past.

File: src/core/layerobserver.cpp [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+  for ( const QgsFeatureId fid : changedAttributesFids )
+    changedFids.insert( fid );
+
+  // NOTE we read the features from the dataProvider directly as we want to access the old values.
+  // If we use the layer, we get the values from the edit buffer.
+  QgsFeatureIterator featuresIt = vl->dataProvider()->getFeatures( QgsFeatureRequest( changedFids ) );
+  QgsChangedFeatures changedFeatures;
+  QgsFeature f;
+
+  // ? is it possible to use the iterator in a less ugly way? something like normal `for ( QgsFeature &f : it ) {}`
```
Comment (m-kuhn) [inline src/core/layerobserver.cpp:115]: No :-(

File: src/core/layerobserver.cpp [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+{
+  const QList<QgsMapLayer *> layers = mProject->mapLayers().values();
+
+  // we should keep track only of the layers on cloud projects
+  if ( QFieldCloudUtils::getProjectId( mProject ).isEmpty() )
+    return;
+
+  for ( QgsMapLayer *layer : layers )
+  {
+    QgsVectorLayer *vl = dynamic_cast<QgsVectorLayer *>( layer );
```
Comment (m-kuhn) [inline src/core/layerobserver.cpp:257]:
```suggestion
    QgsVectorLayer *vl = qobject_cast<QgsVectorLayer *>( layer );
```

File: src/core/deltastatuslistmodel.cpp [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "deltastatuslistmodel.h"
+#include <QJsonObject>
+#include <QJsonArray>
+#include <QDebug>
```
Comment (m-kuhn) [inline src/core/deltastatuslistmodel.cpp:19]:
```suggestion
```

File: src/core/utils/qfieldcloudutils.cpp [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+ ***************************************************************************/
+
+#include "qfieldcloudutils.h"
+#include <qgsapplication.h>
+#include <QDir>
+#include <QString>
+#include "qgsmessagelog.h"
+
+// NOTE directly setting does not work QgsApplication::qgisSettingsDirPath();
+QString QFieldCloudUtils::sQgisSettingsDirPath = QString();
```
[the remark this answers is not in the document]
Comment (suricactus) [reply src/core/utils/qfieldcloudutils.cpp:23]: If I undestood you correctly, no. This is used only for testing purposes here:
https://github.com/opengisch/QField/blob/ba6deb1eefd959453cd714d5179e5298b248e81e/test/test_deltafilewrapper.cpp#L43

**Review by m-kuhn — APPROVED** [review]
image
supergreen

**Comment by nirvn** [comment]
Woupidou!!

**Pull request merged**